orchestration/orchestration.go

```diff
@@ -95,36 +98,50 @@ func (o *Orchestrator) CreateService(ctx context.Context, input *ServiceOrchestr
 	}
 	input.Tags = ct
 
+	// setup err var, rollback function list and defer execution, note that we depend on the err variable defined above this
+	var rollBackTasks []rollbackFunc
+	defer func() {
+		if err != nil {
+			log.Errorf("recovering from error: %s, executing %d rollback tasks", err, len(rollBackTasks))
+			go rollBack(&rollBackTasks)
+		}
+	}()
+
 	output := &ServiceOrchestrationOutput{}
-	cluster, err := o.processCluster(ctx, input)
+	cluster, rbfunc, err := o.processCluster(ctx, input)
 	if err != nil {
 		return nil, err
 	}
 	output.Cluster = cluster
+	rollBackTasks = append(rollBackTasks, rbfunc)
 
-	creds, err := o.processRepositoryCredentials(ctx, input)
+	creds, rbfunc, err := o.processRepositoryCredentials(ctx, input)
```

Comment:
Looks like you're shadowing `err` here and below, making me think the rollback may not work as expected.

Reply:
I think it's okay to shadow err since it will always `return` before the variable is reset.  (I'm not sure it's even considered shadowing since it's in the same scope, not nested?).  If I had something continuing on after catching an error, we might not want to replace it.

Reply:
I guess that should be fine then. I'm still not sure why Go lets you re-declare an existing variable (in the same scope) and what exactly that does under the covers.

Reply:
it replaces it with a new variable (and garbage collects the old one - although i think that happens on function return).  if it was a new scope, it would shadow the outer variable until the scope ended (like an if block)...